Cluster spec docs for EKS-A cluster on CloudStack (#3677)
aws/eks-anywhere: docs/content/en/docs/reference/clusterspec/cloudstack.md

    count: 2
    endpoint:
      host: ""
    machineGroupRef:
      kind: CloudStackMachineConfig
      name: my-cluster-name-cp
    taints:
    - key: ""
      value: ""
      effect: ""
    labels:
      "<key1>": ""
      "<key2>": ""
  datacenterRef:
    kind: CloudStackDatacenterConfig
    name: my-cluster-name
  externalEtcdConfiguration:
    count: 3
    machineGroupRef:
      kind: CloudStackMachineConfig
      name: my-cluster-name-etcd
  kubernetesVersion: "1.23"
  managementCluster:
    name: my-cluster-name
  workerNodeGroupConfigurations:
  - count: 2
    machineGroupRef:
      kind: CloudStackMachineConfig
      name: my-cluster-name
    taints:
    - key: ""
      value: ""
      effect: ""
    labels:
      "<key1>": ""
      "<key2>": ""
---
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: CloudStackDatacenterConfig
metadata:
  name: my-cluster-name-datacenter
spec:
  availabilityZones:
  - account: admin
    credentialsRef: global
    domain: domain1
    managementApiEndpoint: ""
    name: az-1
    zone:
      network: {}
---
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: CloudStackMachineConfig
metadata:
  name: my-cluster-name-cp

> spec:

Author: I added those fields.

  computeOffering: {}

>   template: {}

Author: Done.

You may define one or more worker node groups.

### workerNodeGroupConfigurations.count (required)
Number of worker nodes

### workerNodeGroupConfigurations.machineGroupRef (required)
Refers to the Kubernetes object with CloudStack specific configuration for your nodes. See `CloudStackMachineConfig Fields` below.

### workerNodeGroupConfigurations.name (required)
Name of the worker node group (default: md-0)

### workerNodeGroupConfigurations.taints
A list of taints to apply to the nodes in the worker node group.

Modifying the taints associated with a worker node group configuration will cause new nodes to be rolled-out, replacing the existing nodes associated with the configuration.

At least one node group must not have `NoSchedule` or `NoExecute` taints applied to it.

### workerNodeGroupConfigurations.labels
A list of labels to apply to the nodes in the worker node group. This is in addition to the labels that
EKS Anywhere will add by default.

Modifying the labels associated with a worker node group configuration will cause new nodes to be rolled out, replacing
the existing nodes associated with the configuration.

## CloudStackDatacenterConfig

> ### availabilityZones.account (required)

Author: Okay. I made it optional and pointed to the description of adding the credentials in the getting started guide. Let me know if anything further is needed.

Account used to access CloudStack. The default is `admin`.

> ### availabilityZones.credentialsRef (required)

Author: Done.

### availabilityZones.domain (required)
CloudStack domain to deploy the cluster. The default is `ROOT`.

### availabilityZones.managementApiEndpoint (required)
Location of the CloudStack API management endpoint. For example, `http://10.11.0.2:8080/client/api`.

### availabilityZones.name (required)
Name of the CloudStack zone on which to deploy the cluster.

> ### availabilityZones.zone.network (required)

Reviewer: From the description, should this be rewritten as
```suggestion
### availabilityZones.zone.name (required)
```

Note that a user can also pass in `availabilityZones.zone.id` instead.

There should also be a separate line item for `availabilityZones.zone` which can either contain a name or an id

Author: I think I get what you are saying. Please check once I commit this.

CloudStack network name to use with the cluster.

## CloudStackMachineConfig
In the example above, there are separate `CloudStackMachineConfig` sections for the control plane (`my-cluster-name-cp`), worker (`my-cluster-name`) and etcd (`my-cluster-name-etcd`)  nodes.

> ### computeOfferings

Author: Done

> Name of the CloudStack compute instance.

Author: Done.

### users[0].name (optional)
The name of the user you want to configure to access your virtual machines through ssh.

The default is `capc`.

### users[0].sshAuthorizedKeys (optional)
The SSH public keys you want to configure to access your virtual machines through ssh (as described below). Only 1 is supported at this time.

### users[0].sshAuthorizedKeys[0] (optional)
This is the SSH public key that will be placed in `authorized_keys` on all EKS Anywhere cluster VMs so you can ssh into
them. The user will be what is defined under name above. For example:

```
ssh -i <private-key-file> <user>@<VM-IP>
```

The default is generating a key in your `$(pwd)/<cluster-name>` folder when not specifying a value.

> ### template (optional)

Author: Done.

> The VM template to use for your EKS Anywhere cluster. Currently, a VM based on RHEL 8.4 is required.

Reviewer: ```suggestion
The VM template to use for your EKS Anywhere cluster. Currently, a VM based on RHEL 8.6 is required.
```

Author: Done

description: >
  Full EKS Anywhere configuration reference for a CloudStack cluster.
---
This is a generic template with detailed descriptions below for reference.
The following additional optional configuration can also be included:

* [CNI]({{< relref "optional/cni.md" >}})
* [IAM for pods]({{< relref "optional/irsa.md" >}})
* [IAM Authenticator]({{< relref "optional/iamauth.md" >}})
* [OIDC]({{< relref "optional/oidc.md" >}})
* [gitops]({{< relref "optional/gitops.md" >}})
* [proxy]({{< relref "optional/proxy.md" >}})
* [Registry Mirror]({{< relref "optional/registrymirror.md" >}})

```yaml
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: Cluster
metadata:
  name: my-cluster-name
spec:
  clusterNetwork:
    cniConfig:
      cilium: {}
    pods:
      cidrBlocks:
      - 192.168.0.0/16
    services:
      cidrBlocks:
      - 10.96.0.0/12
  controlPlaneConfiguration:
    count: 3
    endpoint:
      host: ""
    machineGroupRef:
      kind: CloudStackMachineConfig
      name: my-cluster-name-cp
    taints:
    - key: ""
      value: ""
      effect: ""
    labels:
      "<key1>": ""

>       "<key2>": ""

Reviewer 2: A special label value is supported by CAPC provider:

```
labels:
      cluster.x-k8s.io/failure-domain: ds.meta_data.failuredomain
```

`ds.meta_data.failuredomain` will be replaced with a failuredomain name where the node is deployed in, for example `az-1`

Author: I like that wording. Done.

  datacenterRef:
    kind: CloudStackDatacenterConfig
    name: my-cluster-name
  externalEtcdConfiguration:
    count: 3
    machineGroupRef:
      kind: CloudStackMachineConfig
      name: my-cluster-name-etcd
  kubernetesVersion: "1.23"
  managementCluster:
    name: my-cluster-name
  workerNodeGroupConfigurations:
  - count: 2
    machineGroupRef:
      kind: CloudStackMachineConfig
      name: my-cluster-name
    taints:
    - key: ""
      value: ""
      effect: ""
    labels:
      "<key1>": ""

>       "<key2>": ""

Reviewer 2: A special label value is supported by CAPC provider: 

```
labels:
      cluster.x-k8s.io/failure-domain: ds.meta_data.failuredomain
```
`ds.meta_data.failuredomain` will be replaced with a failuredomain name where the node is deployed in, for example `az-1`

Author: Done.

---
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: CloudStackDatacenterConfig
metadata:
  name: my-cluster-name-datacenter
spec:
  availabilityZones:
  - account: admin
    credentialsRef: global
    domain: domain1
    managementApiEndpoint: ""
    name: az-1
    zone:
      name: zone1
      network:
        name: "net1"
---
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: CloudStackMachineConfig
metadata:
  name: my-cluster-name-cp
spec:
  computeOffering:
    name: "m4-large"
  users:
  - name: capc
    sshAuthorizedKeys:
    - ssh-rsa AAAA...
  template:
    name: "rhel8-k8s-118"
  diskOffering:
    name: "Small"
    mountPath: "/data-small"
    device: "/dev/vdb"

>     filesystem" "ext4"

Author: Done.

### managementCluster (required)
Identifies the name of the management cluster.
If this is a standalone cluster or if it were serving as the management cluster for other workload clusters, this will be the same as the cluster name.

### workerNodeGroupConfigurations (required)
This takes in a list of node groups that you can define for your workers.
You may define one or more worker node groups.

### workerNodeGroupConfigurations.count (required)
Number of worker nodes

### workerNodeGroupConfigurations.machineGroupRef (required)
Refers to the Kubernetes object with CloudStack specific configuration for your nodes. See `CloudStackMachineConfig Fields` below.

### workerNodeGroupConfigurations.name (required)
Name of the worker node group (default: md-0)

### workerNodeGroupConfigurations.taints
A list of taints to apply to the nodes in the worker node group.

Modifying the taints associated with a worker node group configuration will cause new nodes to be rolled-out, replacing the existing nodes associated with the configuration.

At least one node group must not have `NoSchedule` or `NoExecute` taints applied to it.

### workerNodeGroupConfigurations.labels

> A list of labels to apply to the nodes in the worker node group. This is in addition to the labels that

Reviewer: There's a special contract here between EKS-A and CAPC. If the user provides the label here `ds.meta_data.failuredomain` then CAPC will substitute the value to provide the actual AZ name as a label on the worker nodes. @wanyufe has the lower level details if you have questions

Author: Okay. I talked to @wanyufe about this. I added information here and to the control plane labels. Please check that I got the wording right.

### workerNodeGroupConfigurations (required)
This takes in a list of node groups that you can define for your workers.
You may define one or more worker node groups.

### workerNodeGroupConfigurations.count (required)
Number of worker nodes

### workerNodeGroupConfigurations.machineGroupRef (required)
Refers to the Kubernetes object with CloudStack specific configuration for your nodes. See `CloudStackMachineConfig Fields` below.

### workerNodeGroupConfigurations.name (required)
Name of the worker node group (default: md-0)

### workerNodeGroupConfigurations.taints
A list of taints to apply to the nodes in the worker node group.

Modifying the taints associated with a worker node group configuration will cause new nodes to be rolled-out, replacing the existing nodes associated with the configuration.

At least one node group must not have `NoSchedule` or `NoExecute` taints applied to it.

### workerNodeGroupConfigurations.labels
A list of labels to apply to the nodes in the worker node group. This is in addition to the labels that
EKS Anywhere will add by default.

> For example, this label identifies a Cluster API failuredomain, which causes CAPC to substitute the value for the actual AZ name as a label on the nodes:

Reviewer: I'm not sure if "for example" makes sense here, since this keyword `ds.meta_data.failuredomain` is specifically reserved for the availability zones replacement usecase. It's more of a custom feature than a generic example

Author: Okay. I think it's the only custom label as well. I'll remove "For example".

Author: I reworded base on comments above.

Reviewer 2: Customer can add other labels like `foo: bar` and nodes will be labeled like that.

`ds.meta_data.failuredomain` is a special keyword case Cloudstack supported.

```yaml
    labels:
      cluster.x-k8s.io/failure-domain: ds.meta_data.failuredomain
```

Modifying the labels associated with a worker node group configuration will cause new nodes to be rolled out, replacing
the existing nodes associated with the configuration.

## CloudStackDatacenterConfig

### availabilityZones.account (optional)
Account used to access CloudStack.

> The default is `admin`.

Reviewer: I'm actually not sure if the default is admin. I think the default would be a lookup to whatever user the api key and secret key were generated from. I'd remove the point about defaulting.

Author: Removed.

Reviewer 2: If account is not specified and there are more than one account under the domain, CAPC will throw an error. If there is only one account, it will use it -- it may not be `admin`

As long as you pass valid credentials, through `availabilityZones.credentialsRef`, this value is not required.

> ### availabilityZones.credentialsRef (optional)

Reviewer: This is required. You can specify either the profile name (from ini file) or the previously created secret on the cluster, in the `eksa-system` namespace

Author: Done.

If you passed credentials through the environment variable `EKSA_CLOUDSTACK_B64ENCODED_SECRET` noted in [Create CloudStack production cluster]({{< relref "../../getting-started/production-environment/cloudstack.md" >}}), you can identify those credentials here.
For that example, you would use the value `global`.
You can also use a previously created Kubernetes secret.

> ### availabilityZones.domain (required)

Reviewer: this is optional, it should default to ROOT

Author: Got it. Thanks.

CloudStack domain to deploy the cluster. The default is `ROOT`.

### availabilityZones.managementApiEndpoint (required)
Location of the CloudStack API management endpoint. For example, `http://10.11.0.2:8080/client/api`.

> ### availabilityZones.name (required)
> Name of the CloudStack zone on which to deploy the cluster.
> As an alternative, you could pass a Zone ID to `availabilityZones.id`.

Reviewer: ```suggestion
### availabilityZones.name (required)
Unique name to categorize the availability zone in which the cluster will reside
```

zone name/id are specified at availabilityZones.zone.{id,name}

> ### availabilityZones.zone.network.name (required)

Reviewer: I don't think network *name* is required specifically. You have to specify a network. To do that, you can _either_ provide `name` or `id`. Similar to `computeOffering` below

Author: Done.

CloudStack network name to use with the cluster.

## CloudStackMachineConfig
In the example above, there are separate `CloudStackMachineConfig` sections for the control plane (`my-cluster-name-cp`), worker (`my-cluster-name`) and etcd (`my-cluster-name-etcd`) nodes.

> ### computeOfferings (required)

Reviewer: ```suggestion
### computeOffering (required)
```

Name or ID of the CloudStack compute instance.

> ### users[0].name (optional)

Reviewer: I think you can pass in an arbitrary number of user objects, not just 1

Author: Done.

The name of the user you want to configure to access your virtual machines through ssh.

The default is `capc`.

### users[0].sshAuthorizedKeys (optional)
The SSH public keys you want to configure to access your virtual machines through ssh (as described below). Only 1 is supported at this time.

### users[0].sshAuthorizedKeys[0] (optional)
This is the SSH public key that will be placed in `authorized_keys` on all EKS Anywhere cluster VMs so you can ssh into
them. The user will be what is defined under name above. For example:

```
ssh -i <private-key-file> <user>@<VM-IP>
```

The default is generating a key in your `$(pwd)/<cluster-name>` folder when not specifying a value.

### template (required)
The VM template to use for your EKS Anywhere cluster. Currently, a VM based on RHEL 8.6 is required.

> This can be a name or ID.

Reviewer: to be clear, it would look like
```
template:
  name: xxxx
```

or 

```
template:
  id: yyyy
```

so one of template.name or template.id are required. Same for a lot of the other fields that reference CloudStack resources

Author: I used the {id,name} notation on each field it applies to. I'll go through later and see if I missed any.

See the [Artifacts]({{< relref "../artifacts" >}}) page for instructions for building RHEL-based images.

> ### diskOffering (optional)

Reviewer: diskOffering.name?

Author: Done.

Name representing a disk you want to mount into nodes for this CloudStackMachineConfig

> ### mountPath (optional)

Reviewer: diskOffering.mountPath? All the diskoffering attributes should be nested under diskOffering

Author: Done.

Mount point on which to mount the disk.

### device (optional)
Device name of the disk partition to mount.

### filesystem (optional)
File system type used to format the filesystem on the disk.

### label (optional)
Label to apply to the disk partition.

> 

Reviewer: could you add the optional `userCustomDetails` field as well?

Author: I added it. It looks like a way of adding key/value pairs to nodes in a CloudStackMachineConfig. Is there some purpose I could give as an example?

Author: Got it. Thanks.

### symlinks (optional)
Symbolic link of a directory or file you want to mount from the host filesystem to the mounted filesystem.

> ### affinityGroupIDs (optional)

Reviewer: I think we're also missing the optional field `affinity` which enables a user to set the `pro` or `anti`, and create and manage an affinity group on their behalf. The `affinity` field can be used in a mutually exclusive fashion with the `affinityGroupIDs` field

Author: Not sure I got this right. Let's revisit after everything else is done.
